plutus-ledger-api/test/Spec/Value.hs

```diff
@@ -0,0 +1,231 @@
+-- list comprises all possible versions of the head that we get by removing any number of
+-- 'Nothing's.
+--
+-- Rendering 'Nothing' as @*@ and @Just c@ as @c@ we get:
+--
+-- >>> map (map $ maybe '*' id) $ insertHooks "abcd"
+-- ["*ab*cd*","ab*cd*","*ab*cd","ab*cd","*abcd*","abcd*","*abcd","abcd"]
+insertHooks :: [a] -> [[Maybe a]]
+insertHooks xs0 = do
+    -- The fast and slow pointers trick to find the middle of the list. Check out
+    -- https://medium.com/@arifimran5/fast-and-slow-pointer-pattern-in-linked-list-43647869ac99
+    -- if you're not familiar with the idea.
+    let go (_ : _ : xsFast) (x : xsSlow) = do
+            xs' <- go xsFast xsSlow
+            [Just x : xs']
+        go _ xsSlow = do
+            prefix <- [[Nothing], []]
+            suffix <- [[Nothing], []]
+            [prefix ++ map Just xsSlow ++ suffix]
+    xs0' <- go xs0 xs0
+    [Nothing : xs0', xs0']
+
+{-# INLINEABLE currencyLongListOptions #-}
+-- | The last and the biggest list of currencies from 'currencyListOptions' with 'longCurrencyChunk'
+-- inserted in it in various ways as per 'insertHooks'.
+currencyLongListOptions :: [[(CurrencySymbol, [(TokenName, Integer)])]]
+currencyLongListOptions =
+    insertHooks (ListTx.last currencyListOptions) <&> \currencyListWithHooks ->
+    ListTx.concatMap (maybe longCurrencyChunk pure) currencyListWithHooks
+
+listsToValue :: [(CurrencySymbol, [(TokenName, Integer)])] -> Value
+listsToValue = Value . AssocMap.fromList . ListTx.map (fmap AssocMap.fromList)
+
+valueToLists :: Value -> [(CurrencySymbol, [(TokenName, Integer)])]
+valueToLists = ListTx.map (fmap AssocMap.toList) . AssocMap.toList . getValue
+
+-- | Check equality of two compiled 'Value's through UPLC evaluation and annotate the result with
+-- the cost of evaluation.
+eqValueCode :: CompiledCode Value -> CompiledCode Value -> (Bool, PLC.CountingSt)
+eqValueCode valueCode1 valueCode2 = (res, cost) where
+    prog =
+        $$(compile [|| \value1 value2 -> toBuiltin ((value1 :: Value) == value2) ||])
+            `unsafeApplyCode` valueCode1 `unsafeApplyCode` valueCode2
+    (errOrRes, cost)
+        = PLC.runCekNoEmit PLC.defaultCekParameters PLC.counting
+        . PLC.runQuote
+        . PLC.unDeBruijnTermWith (Haskell.error "Free variable")
+        . PLC._progTerm
+        $ getPlc prog
+    res = either Haskell.throw id $ errOrRes >>= PLC.readKnownSelf
+
+-- | Check equality of two compiled 'Value's directly in Haskell.
+haskellEqValue :: Value -> Value -> Bool
+haskellEqValue value1 value2 = toMap value1 Haskell.== toMap value2 where
+    toMap
+        = Map.filter (Haskell.not . Map.null)
+        . Haskell.fmap (Map.filter (Haskell./= 0))
+        . Map.fromListWith (Map.unionWith (Haskell.+))
+        . Haskell.map (Haskell.fmap $ Map.fromListWith (Haskell.+))
+        . valueToLists
+
+-- | Check whether all currencies and tokens within each of the currencies occur uniquely.
+allDistinct :: Value -> Bool
```

Comment:
Can't you use your `safeFromList` to avoid needing this?

Reply:
Maybe, but the point is that it's for tests and I want an alternative implementation for testing purposes that does not rely on anything from `PlutusTx.AssocMap` and is self-contained etc.